src/compiler/crystal/macros/methods.cr

```diff
@@ -746,13 +746,34 @@ module Crystal
           BoolLiteral.new(@value.ends_with?(piece))
         end
       when "gsub"
-        interpret_check_args do |first, second|
-          raise "first argument to StringLiteral#gsub must be a regex, not #{first.class_desc}" unless first.is_a?(RegexLiteral)
-          raise "second argument to StringLiteral#gsub must be a string, not #{second.class_desc}" unless second.is_a?(StringLiteral)
+        if block
+          interpret_check_args(uses_block: true) do |first|
+            raise "first argument to StringLiteral#gsub(&) must be a regex, not #{first.class_desc}" unless first.is_a?(RegexLiteral)
+
+            regex = regex_value first
+
+            new_value = value.gsub regex do |_, match|
+              block.args.each_with_index do |arg, idx|
```

Comment:
Could also consider raising if `block.args.size != match.group_size`. Currently at the moment, the extra block args would just always be `nil`.